<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>Transcriptome Analysis in R</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      {   }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <link rel="stylesheet" href="avenir-white.css" />
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Transcriptome Analysis in R</h1>
</header>
<p>This week we will work with a RNA sequencing data set to analyze patterns of differential expression of genes from human X chromosome. This data sets is described in this publication: <a href="data/Pertea_rnaseq_2016.pdf">Pertea et al (2016)</a>.</p>
<p>The data set itself can be downloaded here (<strong>but you do not need to</strong>): <a href="ftp://ftp.ccb.jhu.edu/pub/RNAseq_protocol/chrX_data.tar.gz">chrX_data.tar.gz</a></p>
<p><br><br></p>
<h2 id="table-of-contents">Table of Contents</h2>
<ul>
<li><p><a href="#typical-rnaseq-analysis-pipeline">1. Typical RNAseq Analysis Pipeline</a></p></li>
<li><p><a href="#data-access">2. Data Access</a></p></li>
<li><p><a href="#data-cleanup">3. Data Cleanup</a></p></li>
<li><p><a href="#r-package-installation">4. Package Installation</a></p></li>
<li><p><a href="#import-data-into-edger">5. Import gene count data in EdgeR</a></p></li>
<li><p><a href="#filter-for-low-expressing-genes">6. Filter For Low Expressing Genes</a></p></li>
<li><p><a href="#normalize-expression-counts-across-samples">7. Normalize Expression Counts Across Samples</a></p></li>
<li><p><a href="#estimate-dispersion">8. Estimate Dispersion</a></p></li>
<li><p><a href="#pairwise-comparison-between-experimental-conditions">9. Pairwise Comparison Between Experimental Conditions</a></p>
<ul>
<li><p><a href="#gbr-female-vs-gbr-male">9.1 GBR Female vs GBR Male</a></p></li>
<li><p><a href="#gbr-female-vs-yri-female">9.2 GBR Female vs YRI Female</a></p></li>
<li><p><a href="#gbr-male-vs-yri-female">9.3 GBR Male vs YRI Female</a></p></li>
<li><p><a href="#gbr-male-vs-yri-male">9.4 GBR Male vs YRI Male</a></p></li>
<li><p><a href="#gbr-female-vs-yri-male">9.5 GBR Female vs YRI Male</a></p></li>
<li><p><a href="#yri-female-vs-yri-male">9.6 YRI Female vs YRI Male</a></p></li>
</ul></li>
<li><p><a href="#subset-all-comparisons-for-the-7-significant-de-genes">10. Subset All Comparisons for the 7 Significant DE Genes</a></p></li>
<li><p><a href="#make-a-heatmap-of-logfc">11. Make A Heatmap of logFC</a></p></li>
<li><p><a href="#homework">12. Homework</a></p></li>
</ul>
<p><br><br><br><br><br><br><br><br> <br><br><br><br><br><br><br><br> <br><br><br><br><br><br><br><br> <br><br><br><br><br><br><br><br> <br><br><br><br><br><br><br><br></p>
<h2 id="typical-rnaseq-analysis-pipeline">1. Typical RNAseq Analysis Pipeline</h2>
<p>Transcriptomic data from short-read sequencing technologies (e.g. Illumina HiSeq) is analyzed in the following manner:</p>
<ol type="1">
<li><p>Obtain sequencing reads from service provider</p></li>
<li><p>Perform quality control to remove low quality sequencing reads</p></li>
<li><p>Obtain sequence data for a reference genome</p></li>
<li><p>Index the reference genome (so it can be efficiently searched for homologies)</p></li>
<li><p>Align post-QC sequencing reads to the indexed reference genome</p></li>
<li><p>Identify transcripts/genes based on their homologies and locations in the reference genome.</p></li>
<li><p>Calculate the transcript abundances in each sample under study and prepare a table of gene and transcript counts.</p></li>
<li><p>Use regression analysis to determine which genes/transcripts are up and down regulated.</p></li>
<li><p>Form hypotheses and test them using the differential expression results.</p></li>
</ol>
<p>Because this is a R tutorial, we will not get into analysis steps 1 through 7. Please consult the publication linked above for the details of those analyses.</p>
<p>We will begin with step #8 above.</p>
<h2 id="data-access">2. Data Access</h2>
<p>In any RNA sequencing experiment, there are two main types of data: genomic and phenotypic. The former comes from sequencing and counting transcripts/genes, and the latter is trait data that will be used to make meaningful inferences from the results.</p>
<h3 id="phenotypic-data">2.1 Phenotypic Data</h3>
<center>
<img src="data/phenotypes.png" width=400>
</center>
<p>The data above can’t be copy pasted. It is an image. We will need to create this data frame in R manually.</p>
<p><br><br></p>
<ol type="1">
<li>First create 3 vectors</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a>sampid &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ERR188245&quot;</span>, <span class="st">&quot;ERR188428&quot;</span>, <span class="st">&quot;ERR188337&quot;</span>, <span class="st">&quot;ERR188401&quot;</span>, <span class="st">&quot;ERR188257&quot;</span>,</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="st">&quot;ERR188383&quot;</span>, <span class="st">&quot;ERR204916&quot;</span>, <span class="st">&quot;ERR188234&quot;</span>, <span class="st">&quot;ERR188273&quot;</span>, <span class="st">&quot;ERR188454&quot;</span>, <span class="st">&quot;ERR188104&quot;</span>,</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="st">&quot;ERR188044&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a>sex &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">replicate</span>(<span class="dv">3</span>, <span class="st">&quot;Female&quot;</span>), <span class="kw">replicate</span>(<span class="dv">3</span>, <span class="st">&quot;Male&quot;</span>), <span class="kw">replicate</span>(<span class="dv">3</span>, <span class="st">&quot;Female&quot;</span>),</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="kw">replicate</span>(<span class="dv">3</span>, <span class="st">&quot;Male&quot;</span>))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a>pop &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">replicate</span>(<span class="dv">6</span>, <span class="st">&quot;GBR&quot;</span>), <span class="kw">replicate</span>(<span class="dv">6</span>, <span class="st">&quot;YRI&quot;</span>))</span></code></pre></div>
<p><br><br></p>
<ol start="2" type="1">
<li>Check to make sure all vectors are of same length</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">length</span>(sampid)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="kw">length</span>(sex)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a><span class="kw">length</span>(pop)</span></code></pre></div>
<p><br><br></p>
<ol start="3" type="1">
<li>Creat a data frame by combining the three vectors as columns:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a>pheno &lt;-<span class="st"> </span><span class="kw">data.frame</span>(sampid, sex, pop)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>pheno</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a>      sampid    sex pop</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a><span class="dv">1</span>  ERR188245 Female GBR</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a><span class="dv">2</span>  ERR188428 Female GBR</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a><span class="dv">3</span>  ERR188337 Female GBR</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a><span class="dv">4</span>  ERR188401   Male GBR</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a><span class="dv">5</span>  ERR188257   Male GBR</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a><span class="dv">6</span>  ERR188383   Male GBR</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a><span class="dv">7</span>  ERR204916 Female YRI</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a><span class="dv">8</span>  ERR188234 Female YRI</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a><span class="dv">9</span>  ERR188273 Female YRI</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a><span class="dv">10</span> ERR188454   Male YRI</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a><span class="dv">11</span> ERR188104   Male YRI</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true"></a><span class="dv">12</span> ERR188044   Male YRI</span></code></pre></div>
<p><br><br></p>
<h3 id="gene-count-data">2.2 Gene Count Data</h3>
<ol type="1">
<li>The gene count data for each sample can be downloaded here: <a href="data/gene_count_matrix.csv">gene_count_matrix.csv</a>. Once downloaded, import the data into R:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>gc &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;gene_count_matrix.csv&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>)</span></code></pre></div>
<ol start="2" type="1">
<li>What are the dimensions of this data frame?</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="kw">dim</span>(gc)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">2082</span>   <span class="dv">13</span></span></code></pre></div>
<ol start="3" type="1">
<li>Check sample names to make sure they match the <code>sampid</code> from the <code>pheno</code> data you created above.</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="kw">names</span>(gc)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a> [<span class="dv">1</span>] <span class="st">&quot;gene_id&quot;</span>   <span class="st">&quot;ERR188044&quot;</span> <span class="st">&quot;ERR188104&quot;</span> <span class="st">&quot;ERR188234&quot;</span> <span class="st">&quot;ERR188245&quot;</span> <span class="st">&quot;ERR188257&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a> [<span class="dv">7</span>] <span class="st">&quot;ERR188273&quot;</span> <span class="st">&quot;ERR188337&quot;</span> <span class="st">&quot;ERR188383&quot;</span> <span class="st">&quot;ERR188401&quot;</span> <span class="st">&quot;ERR188428&quot;</span> <span class="st">&quot;ERR188454&quot;</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a>[<span class="dv">13</span>] <span class="st">&quot;ERR204916&quot;</span></span></code></pre></div>
<ul>
<li><p>Did you notice that the sample names are not in the same order in <code>gc</code> compared to <code>pheno</code>. This is problematic. The analytical software we are going to use requires that both files should have the same order for sample names. How do we fix this?</p></li>
<li><p>We can either change the order in one or the other file. In the pheno file, the order is appropriate for the analysis, which clearly shows replicates for each experimental condition. So we will keep that file as is and change <code>gc</code> instead.</p></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a>gc2 &lt;-<span class="st"> </span>gc[,<span class="kw">c</span>(<span class="st">&quot;gene_id&quot;</span>, sampid)]</span></code></pre></div>
<ul>
<li>Note what did above. We already have the correct order of samples in the vector <code>sampid</code> we created earlier. So we simply substituted it and simply pre-pended <code>gene_id</code> to that list. Check to make sure that the new data frame shows the data correctly:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="kw">head</span>(gc)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a>               gene_id ERR188044 ERR188104 ERR188234 ERR188245 ERR188257 ERR188273 ERR188337 ERR188383 ERR188401 ERR188428 ERR188454 ERR204916</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a><span class="dv">1</span>  NM_<span class="dv">001177465</span><span class="op">|</span>MAMLD1         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a><span class="dv">2</span>      NR_<span class="dv">029427</span><span class="op">|</span>WDR13         <span class="dv">0</span>         <span class="dv">0</span>        <span class="dv">54</span>       <span class="dv">211</span>       <span class="dv">204</span>        <span class="dv">47</span>       <span class="dv">177</span>       <span class="dv">165</span>       <span class="dv">177</span>       <span class="dv">112</span>       <span class="dv">169</span>       <span class="dv">191</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a><span class="dv">3</span>   NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">401</span>       <span class="dv">495</span>       <span class="dv">574</span>       <span class="dv">520</span>       <span class="dv">614</span>        <span class="dv">83</span>       <span class="dv">737</span>       <span class="dv">304</span>       <span class="dv">737</span>       <span class="dv">313</span>       <span class="dv">358</span>       <span class="dv">312</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a><span class="dv">4</span>   NM_<span class="dv">001303516</span><span class="op">|</span>PDZD4         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true"></a><span class="dv">5</span> NM_<span class="dv">001170757</span><span class="op">|</span>FAM122B       <span class="dv">865</span>       <span class="dv">470</span>       <span class="dv">462</span>         <span class="dv">0</span>       <span class="dv">271</span>        <span class="dv">94</span>       <span class="dv">342</span>       <span class="dv">295</span>       <span class="dv">342</span>       <span class="dv">306</span>       <span class="dv">403</span>       <span class="dv">127</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true"></a><span class="dv">6</span>    NM_<span class="dv">001185081</span><span class="op">|</span>FMR1         <span class="dv">0</span>        <span class="dv">95</span>         <span class="dv">0</span>         <span class="dv">0</span>       <span class="dv">346</span>         <span class="dv">0</span>        <span class="dv">82</span>       <span class="dv">294</span>        <span class="dv">82</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true"></a><span class="kw">head</span>(gc2)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true"></a>               gene_id ERR188245 ERR188428 ERR188337 ERR188401 ERR188257 ERR188383 ERR204916 ERR188234 ERR188273 ERR188454 ERR188104 ERR188044</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true"></a><span class="dv">1</span>  NM_<span class="dv">001177465</span><span class="op">|</span>MAMLD1         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true"></a><span class="dv">2</span>      NR_<span class="dv">029427</span><span class="op">|</span>WDR13       <span class="dv">211</span>       <span class="dv">112</span>       <span class="dv">177</span>       <span class="dv">177</span>       <span class="dv">204</span>       <span class="dv">165</span>       <span class="dv">191</span>        <span class="dv">54</span>        <span class="dv">47</span>       <span class="dv">169</span>         <span class="dv">0</span>         <span class="dv">0</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true"></a><span class="dv">3</span>   NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">520</span>       <span class="dv">313</span>       <span class="dv">737</span>       <span class="dv">737</span>       <span class="dv">614</span>       <span class="dv">304</span>       <span class="dv">312</span>       <span class="dv">574</span>        <span class="dv">83</span>       <span class="dv">358</span>       <span class="dv">495</span>       <span class="dv">401</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true"></a><span class="dv">4</span>   NM_<span class="dv">001303516</span><span class="op">|</span>PDZD4         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true"></a><span class="dv">5</span> NM_<span class="dv">001170757</span><span class="op">|</span>FAM122B         <span class="dv">0</span>       <span class="dv">306</span>       <span class="dv">342</span>       <span class="dv">342</span>       <span class="dv">271</span>       <span class="dv">295</span>       <span class="dv">127</span>       <span class="dv">462</span>        <span class="dv">94</span>       <span class="dv">403</span>       <span class="dv">470</span>       <span class="dv">865</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true"></a><span class="dv">6</span>    NM_<span class="dv">001185081</span><span class="op">|</span>FMR1         <span class="dv">0</span>         <span class="dv">0</span>        <span class="dv">82</span>        <span class="dv">82</span>       <span class="dv">346</span>       <span class="dv">294</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>         <span class="dv">0</span>        <span class="dv">95</span>         <span class="dv">0</span></span></code></pre></div>
<ul>
<li>At this point our data sets should be ready for further steps.</li>
</ul>
<p><br><br></p>
<h2 id="data-cleanup">3. Data Cleanup</h2>
<ul>
<li>One thing you may have noticed is that we seem to have many genes where there is no data for any of the 12 samples. These are the genes which read <code>0</code> in each of the 12 columns. There is essentially no data here, so we need to get rid of those rows.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a>gc3 &lt;-<span class="st"> </span>gc2[<span class="kw">apply</span>(gc2[<span class="kw">c</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">13</span>)], <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="op">!</span><span class="kw">any</span>(x <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)),]</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a><span class="kw">dim</span>(gc3)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">656</span>  <span class="dv">13</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a><span class="kw">head</span>(gc3)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true"></a>              gene_id ERR188245 ERR188428 ERR188337 ERR188401 ERR188257 ERR188383 ERR204916 ERR188234 ERR188273 ERR188454 ERR188104 ERR188044</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true"></a><span class="dv">3</span>  NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">520</span>       <span class="dv">313</span>       <span class="dv">737</span>       <span class="dv">737</span>       <span class="dv">614</span>       <span class="dv">304</span>       <span class="dv">312</span>       <span class="dv">574</span>        <span class="dv">83</span>       <span class="dv">358</span>       <span class="dv">495</span>       <span class="dv">401</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true"></a><span class="dv">7</span>     NM_<span class="dv">001356</span><span class="op">|</span>DDX3X       <span class="dv">203</span>       <span class="dv">336</span>       <span class="dv">312</span>       <span class="dv">312</span>       <span class="dv">239</span>        <span class="dv">37</span>       <span class="dv">129</span>       <span class="dv">432</span>        <span class="dv">44</span>        <span class="dv">32</span>       <span class="dv">879</span>       <span class="dv">102</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true"></a><span class="dv">13</span>     NM_<span class="dv">000489</span><span class="op">|</span>ATRX      <span class="dv">1260</span>      <span class="dv">3278</span>      <span class="dv">3919</span>      <span class="dv">3919</span>      <span class="dv">4098</span>      <span class="dv">3632</span>      <span class="dv">4259</span>      <span class="dv">1946</span>      <span class="dv">1692</span>      <span class="dv">2501</span>      <span class="dv">5921</span>      <span class="dv">4909</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true"></a><span class="dv">22</span>      NM_<span class="dv">000116</span><span class="op">|</span>TAZ       <span class="dv">207</span>       <span class="dv">282</span>       <span class="dv">562</span>       <span class="dv">562</span>       <span class="dv">286</span>       <span class="dv">389</span>       <span class="dv">468</span>       <span class="dv">598</span>       <span class="dv">130</span>       <span class="dv">263</span>       <span class="dv">430</span>       <span class="dv">432</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true"></a><span class="dv">23</span>    NM_<span class="dv">016024</span><span class="op">|</span>RBMX2      <span class="dv">1072</span>      <span class="dv">1164</span>      <span class="dv">1849</span>      <span class="dv">1849</span>      <span class="dv">1356</span>      <span class="dv">1752</span>      <span class="dv">1430</span>      <span class="dv">1453</span>       <span class="dv">765</span>      <span class="dv">1292</span>      <span class="dv">1843</span>      <span class="dv">1626</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true"></a><span class="dv">28</span>   NR_<span class="dv">029635</span><span class="op">|</span>MIR221         <span class="dv">1</span>         <span class="dv">2</span>         <span class="dv">8</span>         <span class="dv">8</span>         <span class="dv">5</span>         <span class="dv">2</span>         <span class="dv">3</span>         <span class="dv">3</span>         <span class="dv">2</span>         <span class="dv">2</span>        <span class="dv">10</span>         <span class="dv">6</span></span></code></pre></div>
<ul>
<li><p>What we did above is to remove rows containing only 0s for columns 2 through 14 in the data set. But instead of saving it in the same data frame, we sent output to a new frame called <code>gc3</code>. We now have 656 rows of data, which means 1426 rows had no data, so we discarded them.</p></li>
<li><p>Let’s write this data frame to disk</p></li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="kw">write.table</span>(gc3, <span class="st">&quot;gene_count_matrix_final.txt&quot;</span>, <span class="dt">row.names=</span>F, <span class="dt">col.names=</span>T,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a><span class="dt">quote=</span>F, <span class="dt">sep=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span></code></pre></div>
<ul>
<li>Verify that the file was written correctly by opening in a text editor.</li>
</ul>
<p><br><br></p>
<h2 id="r-package-installation">4. R Package Installation</h2>
<p>For this exercise, we will use the Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">EdgeR</a>. click on the link which will take you to the instructions for installing this package.</p>
<p>Bioconductor is a repository for packages used in biological data analysis. It has a different system for installing packages.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;edgeR&quot;</span>)</span></code></pre></div>
<ul>
<li>Load the package</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="kw">library</span>(edgeR)</span></code></pre></div>
<ul>
<li>Package documentation is available <a href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">here</a></li>
</ul>
<p><br><br></p>
<h2 id="import-data-in-edger">5. Import data in EdgeR</h2>
<p>In our current data frame, the gene name appear in their own column. EdgeR prefers to have this data as row names.</p>
<p><br></p>
<h3 id="import-count-data">5.1 Import count data</h3>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a>ct &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;gene_count_matrix_Final.txt&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">row.names=</span><span class="st">&quot;gene_id&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a><span class="kw">head</span>(ct)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a>                   ERR188245 ERR188428 ERR188337 ERR188401 ERR188257 ERR188383 ERR204916 ERR188234 ERR188273 ERR188454 ERR188104 ERR188044</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a>NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">520</span>       <span class="dv">313</span>       <span class="dv">737</span>       <span class="dv">737</span>       <span class="dv">614</span>       <span class="dv">304</span>       <span class="dv">312</span>       <span class="dv">574</span>        <span class="dv">83</span>       <span class="dv">358</span>       <span class="dv">495</span>       <span class="dv">401</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true"></a>NM_<span class="dv">001356</span><span class="op">|</span>DDX3X          <span class="dv">203</span>       <span class="dv">336</span>       <span class="dv">312</span>       <span class="dv">312</span>       <span class="dv">239</span>        <span class="dv">37</span>       <span class="dv">129</span>       <span class="dv">432</span>        <span class="dv">44</span>        <span class="dv">32</span>       <span class="dv">879</span>       <span class="dv">102</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true"></a>NM_<span class="dv">000489</span><span class="op">|</span>ATRX          <span class="dv">1260</span>      <span class="dv">3278</span>      <span class="dv">3919</span>      <span class="dv">3919</span>      <span class="dv">4098</span>      <span class="dv">3632</span>      <span class="dv">4259</span>      <span class="dv">1946</span>      <span class="dv">1692</span>      <span class="dv">2501</span>      <span class="dv">5921</span>      <span class="dv">4909</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true"></a>NM_<span class="dv">000116</span><span class="op">|</span>TAZ            <span class="dv">207</span>       <span class="dv">282</span>       <span class="dv">562</span>       <span class="dv">562</span>       <span class="dv">286</span>       <span class="dv">389</span>       <span class="dv">468</span>       <span class="dv">598</span>       <span class="dv">130</span>       <span class="dv">263</span>       <span class="dv">430</span>       <span class="dv">432</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true"></a>NM_<span class="dv">016024</span><span class="op">|</span>RBMX2         <span class="dv">1072</span>      <span class="dv">1164</span>      <span class="dv">1849</span>      <span class="dv">1849</span>      <span class="dv">1356</span>      <span class="dv">1752</span>      <span class="dv">1430</span>      <span class="dv">1453</span>       <span class="dv">765</span>      <span class="dv">1292</span>      <span class="dv">1843</span>      <span class="dv">1626</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true"></a>NR_<span class="dv">029635</span><span class="op">|</span>MIR221           <span class="dv">1</span>         <span class="dv">2</span>         <span class="dv">8</span>         <span class="dv">8</span>         <span class="dv">5</span>         <span class="dv">2</span>         <span class="dv">3</span>         <span class="dv">3</span>         <span class="dv">2</span>         <span class="dv">2</span>        <span class="dv">10</span>         <span class="dv">6</span></span></code></pre></div>
<ul>
<li>Notice how the <code>gene_id</code> column has now been converted into row names.</li>
</ul>
<p><br></p>
<h3 id="create-factors-to-describe-phenotypic-data">5.2 Create factors to describe phenotypic data</h3>
<ul>
<li>Next, we need to create factors for our phenotypic data. Our phenotypes consist of two sexes from each of the two populations with three replicates from each. So we have four factors.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a>group &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">4</span>))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a>group</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true"></a> [<span class="dv">1</span>] <span class="dv">1</span> <span class="dv">1</span> <span class="dv">1</span> <span class="dv">2</span> <span class="dv">2</span> <span class="dv">2</span> <span class="dv">3</span> <span class="dv">3</span> <span class="dv">3</span> <span class="dv">4</span> <span class="dv">4</span> <span class="dv">4</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true"></a></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true"></a>Levels<span class="op">:</span><span class="st"> </span><span class="dv">1</span> <span class="dv">2</span> <span class="dv">3</span> <span class="dv">4</span></span></code></pre></div>
<p><br></p>
<h3 id="create-differential-gene-expression-list">5.3 Create Differential Gene Expression List</h3>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a>y &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts=</span>ct, <span class="dt">group=</span>group)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true"></a>y</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true"></a>An object of class <span class="st">&quot;DGEList&quot;</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true"></a><span class="op">$</span>counts</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true"></a>                   ERR188245 ERR188428 ERR188337 ERR188401 ERR188257 ERR188383 ERR204916 ERR188234 ERR188273 ERR188454 ERR188104 ERR188044</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true"></a>NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">520</span>       <span class="dv">313</span>       <span class="dv">737</span>       <span class="dv">737</span>       <span class="dv">614</span>       <span class="dv">304</span>       <span class="dv">312</span>       <span class="dv">574</span>        <span class="dv">83</span>       <span class="dv">358</span>       <span class="dv">495</span>       <span class="dv">401</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true"></a>NM_<span class="dv">001356</span><span class="op">|</span>DDX3X          <span class="dv">203</span>       <span class="dv">336</span>       <span class="dv">312</span>       <span class="dv">312</span>       <span class="dv">239</span>        <span class="dv">37</span>       <span class="dv">129</span>       <span class="dv">432</span>        <span class="dv">44</span>        <span class="dv">32</span>       <span class="dv">879</span>       <span class="dv">102</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true"></a>NM_<span class="dv">000489</span><span class="op">|</span>ATRX          <span class="dv">1260</span>      <span class="dv">3278</span>      <span class="dv">3919</span>      <span class="dv">3919</span>      <span class="dv">4098</span>      <span class="dv">3632</span>      <span class="dv">4259</span>      <span class="dv">1946</span>      <span class="dv">1692</span>      <span class="dv">2501</span>      <span class="dv">5921</span>      <span class="dv">4909</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true"></a>NM_<span class="dv">000116</span><span class="op">|</span>TAZ            <span class="dv">207</span>       <span class="dv">282</span>       <span class="dv">562</span>       <span class="dv">562</span>       <span class="dv">286</span>       <span class="dv">389</span>       <span class="dv">468</span>       <span class="dv">598</span>       <span class="dv">130</span>       <span class="dv">263</span>       <span class="dv">430</span>       <span class="dv">432</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true"></a>NM_<span class="dv">016024</span><span class="op">|</span>RBMX2         <span class="dv">1072</span>      <span class="dv">1164</span>      <span class="dv">1849</span>      <span class="dv">1849</span>      <span class="dv">1356</span>      <span class="dv">1752</span>      <span class="dv">1430</span>      <span class="dv">1453</span>       <span class="dv">765</span>      <span class="dv">1292</span>      <span class="dv">1843</span>      <span class="dv">1626</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true"></a><span class="dv">651</span> more rows ...</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true"></a><span class="op">$</span>samples</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true"></a>          group lib.size norm.factors</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true"></a>ERR188245     <span class="dv">1</span>  <span class="dv">1348647</span>            <span class="dv">1</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true"></a>ERR188428     <span class="dv">1</span>  <span class="dv">1344702</span>            <span class="dv">1</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true"></a>ERR188337     <span class="dv">1</span>  <span class="dv">2098602</span>            <span class="dv">1</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true"></a>ERR188401     <span class="dv">2</span>  <span class="dv">2098602</span>            <span class="dv">1</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true"></a>ERR188257     <span class="dv">2</span>  <span class="dv">1501213</span>            <span class="dv">1</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true"></a><span class="dv">7</span> more rows ...</span></code></pre></div>
<p><br><br></p>
<h2 id="filter-for-low-expressing-genes">6. Filter For Low Expressing Genes</h2>
<p>You may recall that earlier in this process we had removed genes which had no data. But there may still be some genes left in the data with very few counts. Since we will be normalizing count data, such low expression genes might skew our estimates. So we will remove them at this stage.</p>
<ul>
<li>First assign a function within EdgeR library called <code>filterByExpr()</code> to a keyword <code>keep</code> for ease of use.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a>keep &lt;-<span class="st"> </span><span class="kw">filterByExpr</span>(y)</span></code></pre></div>
<ul>
<li>Let the filtering proceed. We</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a>yfilt &lt;-<span class="st"> </span>y[keep, , keep.lib.sizes=<span class="ot">FALSE</span>]</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a>yfilt</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a>An object of class <span class="st">&quot;DGEList&quot;</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true"></a><span class="op">$</span>counts</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true"></a>                   ERR188245 ERR188428 ERR188337 ERR188401 ERR188257 ERR188383 ERR204916 ERR188234 ERR188273 ERR188454 ERR188104 ERR188044</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true"></a>NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">520</span>       <span class="dv">313</span>       <span class="dv">737</span>       <span class="dv">737</span>       <span class="dv">614</span>       <span class="dv">304</span>       <span class="dv">312</span>       <span class="dv">574</span>        <span class="dv">83</span>       <span class="dv">358</span>       <span class="dv">495</span>       <span class="dv">401</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true"></a>NM_<span class="dv">001356</span><span class="op">|</span>DDX3X          <span class="dv">203</span>       <span class="dv">336</span>       <span class="dv">312</span>       <span class="dv">312</span>       <span class="dv">239</span>        <span class="dv">37</span>       <span class="dv">129</span>       <span class="dv">432</span>        <span class="dv">44</span>        <span class="dv">32</span>       <span class="dv">879</span>       <span class="dv">102</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true"></a>NM_<span class="dv">000489</span><span class="op">|</span>ATRX          <span class="dv">1260</span>      <span class="dv">3278</span>      <span class="dv">3919</span>      <span class="dv">3919</span>      <span class="dv">4098</span>      <span class="dv">3632</span>      <span class="dv">4259</span>      <span class="dv">1946</span>      <span class="dv">1692</span>      <span class="dv">2501</span>      <span class="dv">5921</span>      <span class="dv">4909</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true"></a>NM_<span class="dv">000116</span><span class="op">|</span>TAZ            <span class="dv">207</span>       <span class="dv">282</span>       <span class="dv">562</span>       <span class="dv">562</span>       <span class="dv">286</span>       <span class="dv">389</span>       <span class="dv">468</span>       <span class="dv">598</span>       <span class="dv">130</span>       <span class="dv">263</span>       <span class="dv">430</span>       <span class="dv">432</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true"></a>NM_<span class="dv">016024</span><span class="op">|</span>RBMX2         <span class="dv">1072</span>      <span class="dv">1164</span>      <span class="dv">1849</span>      <span class="dv">1849</span>      <span class="dv">1356</span>      <span class="dv">1752</span>      <span class="dv">1430</span>      <span class="dv">1453</span>       <span class="dv">765</span>      <span class="dv">1292</span>      <span class="dv">1843</span>      <span class="dv">1626</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true"></a><span class="dv">640</span> more rows ...</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true"></a><span class="op">$</span>samples</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true"></a>          group lib.size norm.factors</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true"></a>ERR188245     <span class="dv">1</span>  <span class="dv">1348614</span>            <span class="dv">1</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true"></a>ERR188428     <span class="dv">1</span>  <span class="dv">1344673</span>            <span class="dv">1</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true"></a>ERR188337     <span class="dv">1</span>  <span class="dv">2098545</span>            <span class="dv">1</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true"></a>ERR188401     <span class="dv">2</span>  <span class="dv">2098545</span>            <span class="dv">1</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true"></a>ERR188257     <span class="dv">2</span>  <span class="dv">1501166</span>            <span class="dv">1</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true"></a><span class="dv">7</span> more rows ...</span></code></pre></div>
<ul>
<li><p>The filtering function as applied here only keeps those genes that have worthwhile counts in a minimum number of samples. This number of samples is computed from the factors we had created earlier. In our case, that number is <code>3</code> - the group size (3 replicates per condition).</p></li>
<li><p>The <code>keep.lib.sizes=FALSE</code> option allows EdgeR to recalculate the library size (total number of reads in each sample) after such low count genes have been removed. The opposite of this would be setting this option to <code>TRUE</code> and may lead to discrepancies.</p></li>
</ul>
<p><br><br></p>
<h2 id="normalize-expression-counts-across-samples">7. Normalize Expression Counts Across Samples</h2>
<ul>
<li>Normalizing libraries between samples is essential to getting accurate downstream results.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a>ynorm &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(yfilt)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a>ynorm<span class="op">$</span>samples</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true"></a>          group lib.size norm.factors</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true"></a>ERR188245     <span class="dv">1</span>  <span class="dv">1348614</span>    <span class="fl">0.9683824</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true"></a>ERR188428     <span class="dv">1</span>  <span class="dv">1344673</span>    <span class="fl">0.9242010</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true"></a>ERR188337     <span class="dv">1</span>  <span class="dv">2098545</span>    <span class="fl">1.0695881</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true"></a>ERR188401     <span class="dv">2</span>  <span class="dv">2098545</span>    <span class="fl">1.0695881</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true"></a>ERR188257     <span class="dv">2</span>  <span class="dv">1501166</span>    <span class="fl">1.0594384</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true"></a>ERR188383     <span class="dv">2</span>  <span class="dv">1530767</span>    <span class="fl">1.0352421</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true"></a>ERR204916     <span class="dv">3</span>  <span class="dv">1697815</span>    <span class="fl">1.0025118</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true"></a>ERR188234     <span class="dv">3</span>  <span class="dv">2419080</span>    <span class="fl">0.9015856</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true"></a>ERR188273     <span class="dv">3</span>   <span class="dv">917517</span>    <span class="fl">0.9375173</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true"></a>ERR188454     <span class="dv">4</span>  <span class="dv">1600884</span>    <span class="fl">1.0658455</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true"></a>ERR188104     <span class="dv">4</span>  <span class="dv">1988270</span>    <span class="fl">1.0055686</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true"></a>ERR188044     <span class="dv">4</span>  <span class="dv">2063117</span>    <span class="fl">0.9805151</span></span></code></pre></div>
<p><br><br></p>
<h2 id="estimate-dispersion">8. Estimate Dispersion</h2>
<ul>
<li>Differential gene expression analysis uses generalized linear models. It becomes necessary to understand how much gene expression varies in different samples from a mean of all samples. That estimate is called dispersion.</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a>ydisp &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(ynorm)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a>ydisp</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true"></a>An object of class <span class="st">&quot;DGEList&quot;</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true"></a><span class="op">$</span>counts</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true"></a>                   ERR188245 ERR188428 ERR188337 ERR188401 ERR188257 ERR188383 ERR204916 ERR188234 ERR188273 ERR188454 ERR188104 ERR188044</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true"></a>NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3       <span class="dv">520</span>       <span class="dv">313</span>       <span class="dv">737</span>       <span class="dv">737</span>       <span class="dv">614</span>       <span class="dv">304</span>       <span class="dv">312</span>       <span class="dv">574</span>        <span class="dv">83</span>       <span class="dv">358</span>       <span class="dv">495</span>       <span class="dv">401</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true"></a>NM_<span class="dv">001356</span><span class="op">|</span>DDX3X          <span class="dv">203</span>       <span class="dv">336</span>       <span class="dv">312</span>       <span class="dv">312</span>       <span class="dv">239</span>        <span class="dv">37</span>       <span class="dv">129</span>       <span class="dv">432</span>        <span class="dv">44</span>        <span class="dv">32</span>       <span class="dv">879</span>       <span class="dv">102</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true"></a>NM_<span class="dv">000489</span><span class="op">|</span>ATRX          <span class="dv">1260</span>      <span class="dv">3278</span>      <span class="dv">3919</span>      <span class="dv">3919</span>      <span class="dv">4098</span>      <span class="dv">3632</span>      <span class="dv">4259</span>      <span class="dv">1946</span>      <span class="dv">1692</span>      <span class="dv">2501</span>      <span class="dv">5921</span>      <span class="dv">4909</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true"></a>NM_<span class="dv">000116</span><span class="op">|</span>TAZ            <span class="dv">207</span>       <span class="dv">282</span>       <span class="dv">562</span>       <span class="dv">562</span>       <span class="dv">286</span>       <span class="dv">389</span>       <span class="dv">468</span>       <span class="dv">598</span>       <span class="dv">130</span>       <span class="dv">263</span>       <span class="dv">430</span>       <span class="dv">432</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true"></a>NM_<span class="dv">016024</span><span class="op">|</span>RBMX2         <span class="dv">1072</span>      <span class="dv">1164</span>      <span class="dv">1849</span>      <span class="dv">1849</span>      <span class="dv">1356</span>      <span class="dv">1752</span>      <span class="dv">1430</span>      <span class="dv">1453</span>       <span class="dv">765</span>      <span class="dv">1292</span>      <span class="dv">1843</span>      <span class="dv">1626</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true"></a><span class="dv">640</span> more rows ...</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true"></a></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true"></a><span class="op">$</span>samples</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true"></a>          group lib.size norm.factors</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true"></a>ERR188245     <span class="dv">1</span>  <span class="dv">1348614</span>    <span class="fl">0.9683824</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true"></a>ERR188428     <span class="dv">1</span>  <span class="dv">1344673</span>    <span class="fl">0.9242010</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true"></a>ERR188337     <span class="dv">1</span>  <span class="dv">2098545</span>    <span class="fl">1.0695881</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true"></a>ERR188401     <span class="dv">2</span>  <span class="dv">2098545</span>    <span class="fl">1.0695881</span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true"></a>ERR188257     <span class="dv">2</span>  <span class="dv">1501166</span>    <span class="fl">1.0594384</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true"></a><span class="dv">7</span> more rows ...</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true"></a></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true"></a><span class="op">$</span>common.dispersion</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="fl">0.226083</span></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true"></a></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true"></a><span class="op">$</span>trended.dispersion</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="fl">0.2841544</span> <span class="fl">0.3252375</span> <span class="fl">0.1064144</span> <span class="fl">0.2970612</span> <span class="fl">0.1550413</span></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true"></a><span class="dv">640</span> more elements ...</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true"></a></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true"></a><span class="op">$</span>tagwise.dispersion</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="fl">0.14545759</span> <span class="fl">0.71584346</span> <span class="fl">0.14771572</span> <span class="fl">0.11178168</span> <span class="fl">0.04818607</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true"></a><span class="dv">640</span> more elements ...</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true"></a></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true"></a><span class="op">$</span>AveLogCPM</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true"></a>[<span class="dv">1</span>]  <span class="fl">8.012388</span>  <span class="fl">7.168844</span> <span class="fl">10.977294</span>  <span class="fl">7.764692</span>  <span class="fl">9.739103</span></span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true"></a><span class="dv">640</span> more elements ...</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true"></a></span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true"></a><span class="op">$</span>trend.method</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">&quot;locfit&quot;</span></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true"></a></span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true"></a><span class="op">$</span>prior.df</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="fl">2.607188</span></span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true"></a></span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true"></a><span class="op">$</span>prior.n</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="fl">0.3258985</span></span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true"></a></span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true"></a><span class="op">$</span>span</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="fl">0.4588173</span></span></code></pre></div>
<p><br><br></p>
<h2 id="pairwise-comparison-between-experimental-conditions">9. Pairwise Comparison Between Experimental Conditions</h2>
<ul>
<li>Finally, we are ready to perform differential gene expression analysis. Let’s take a quick look at our phenotypic data to decide which pair to try first.</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a>pheno</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true"></a>      sampid    sex pop</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true"></a><span class="dv">1</span>  ERR188245 Female GBR</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true"></a><span class="dv">2</span>  ERR188428 Female GBR</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true"></a><span class="dv">3</span>  ERR188337 Female GBR</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true"></a><span class="dv">4</span>  ERR188401   Male GBR</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true"></a><span class="dv">5</span>  ERR188257   Male GBR</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true"></a><span class="dv">6</span>  ERR188383   Male GBR</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true"></a><span class="dv">7</span>  ERR204916 Female YRI</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true"></a><span class="dv">8</span>  ERR188234 Female YRI</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true"></a><span class="dv">9</span>  ERR188273 Female YRI</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true"></a><span class="dv">10</span> ERR188454   Male YRI</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true"></a><span class="dv">11</span> ERR188104   Male YRI</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true"></a><span class="dv">12</span> ERR188044   Male YRI</span></code></pre></div>
<p><br></p>
<h3 id="gbr-female-vs-gbr-male">9.1 GBR Female vs GBR Male</h3>
<ul>
<li>According to our group factors, these are groups 1 and 2. How can we verify this?</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a><span class="kw">levels</span>(ydisp<span class="op">$</span>samples<span class="op">$</span>group)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">&quot;1&quot;</span> <span class="st">&quot;2&quot;</span> <span class="st">&quot;3&quot;</span> <span class="st">&quot;4&quot;</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true"></a><span class="kw">rownames</span>(ydisp<span class="op">$</span>samples)[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">&quot;ERR188245&quot;</span> <span class="st">&quot;ERR188428&quot;</span> <span class="st">&quot;ERR188337&quot;</span> <span class="st">&quot;ERR188401&quot;</span></span></code></pre></div>
<ul>
<li><p>Cross check to make sure that the <code>ERR</code> library names match between the pheno files for the phenotypes we are selecting.</p></li>
<li><p>Then use Fisher’s Exact test to estimate DEGs for this pair of samples.</p></li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true"></a>GBR_MF &lt;-<span class="st"> </span><span class="kw">exactTest</span>(ydisp, <span class="dt">pair=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true"></a>GBR_MF</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true"></a>An object of class <span class="st">&quot;DGEExact&quot;</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true"></a><span class="op">$</span>table</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true"></a>                         logFC    logCPM    PValue</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true"></a>NM_<span class="dv">001242640</span><span class="op">|</span>BRCC3 <span class="fl">-0.11039010</span>  <span class="fl">8.012388</span> <span class="fl">0.8097677</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true"></a>NM_<span class="dv">001356</span><span class="op">|</span>DDX3X    <span class="fl">-0.85210878</span>  <span class="fl">7.168844</span> <span class="fl">0.4124643</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true"></a>NM_<span class="dv">000489</span><span class="op">|</span>ATRX      <span class="fl">0.30650253</span> <span class="fl">10.977294</span> <span class="fl">0.5020825</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true"></a>NM_<span class="dv">000116</span><span class="op">|</span>TAZ       <span class="fl">0.08702188</span>  <span class="fl">7.764692</span> <span class="fl">0.8299446</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true"></a>NM_<span class="dv">016024</span><span class="op">|</span>RBMX2     <span class="fl">0.10799335</span>  <span class="fl">9.739103</span> <span class="fl">0.6799526</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true"></a><span class="dv">640</span> more rows ...</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true"></a><span class="op">$</span>comparison</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">&quot;1&quot;</span> <span class="st">&quot;2&quot;</span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true"></a></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true"></a><span class="op">$</span>genes</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true"></a><span class="ot">NULL</span></span></code></pre></div>
<ul>
<li><p>You will notice in the table that we have three types of values.</p>
<ul>
<li><code>logFC</code> is log fold change, a metric of how different gene expression is between the samples</li>
<li><code>logCPM</code> is the log of counts per million (you can disregard this for now)</li>
<li><code>PValue</code> is your standard measure of significance from Fisher’s exact test.</li>
</ul></li>
<li><p>Doing multiple testing and simply using P values for inference may lead to spurious false positives and/or false negatives.</p></li>
<li><p>To make our results more robust, we will estimate <code>FDR</code> or False Discovery Rate which is preferable to using Pvalues.</p></li>
</ul>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true"></a>GBR_MF_FDR &lt;-<span class="st"> </span><span class="kw">topTags</span>(GBR_MF, <span class="dt">n=</span><span class="ot">Inf</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true"></a><span class="kw">head</span>(GBR_MF_FDR)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true"></a>Comparison of groups<span class="op">:</span><span class="st">  </span><span class="dv">2-1</span> </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true"></a>                        logFC    logCPM       PValue        FDR</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true"></a>NR_<span class="dv">003255</span><span class="op">|</span>TSIX     <span class="fl">-5.3282667</span>  <span class="fl">8.731971</span> <span class="fl">4.682312e-05</span> <span class="fl">0.03020091</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true"></a>NM_<span class="dv">004187</span><span class="op">|</span>KDM5C    <span class="fl">-0.6818842</span> <span class="fl">10.926060</span> <span class="fl">1.801629e-02</span> <span class="fl">1.00000000</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true"></a>NR_<span class="dv">132647</span><span class="op">|</span>AIFM1     <span class="fl">1.6119246</span>  <span class="fl">5.374244</span> <span class="fl">2.130455e-02</span> <span class="fl">1.00000000</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true"></a>NM_<span class="dv">001282622</span><span class="op">|</span>KDM5C <span class="fl">-0.6675069</span> <span class="fl">10.949292</span> <span class="fl">2.165767e-02</span> <span class="fl">1.00000000</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true"></a>NR_<span class="dv">033699</span><span class="op">|</span>FMR1      <span class="fl">1.3293297</span>  <span class="fl">7.244091</span> <span class="fl">2.210837e-02</span> <span class="fl">1.00000000</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true"></a>NM_<span class="dv">001281463</span><span class="op">|</span>SMC1A  <span class="fl">4.2982844</span>  <span class="fl">8.333749</span> <span class="fl">2.417667e-02</span> <span class="fl">1.00000000</span></span></code></pre></div>
<ul>
<li>Write this data to a file</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true"></a><span class="kw">write.table</span>(GBR_MF_FDR, <span class="st">&quot;GBR_MF_FDR.txt&quot;</span>, <span class="dt">row.names=</span>T, <span class="dt">col.names=</span>T, <span class="dt">quote=</span>F, <span class="dt">sep=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span></code></pre></div>
<ul>
<li>Which genes are significant for DE?</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true"></a><span class="kw">sum</span>(GBR_MF_FDR<span class="op">$</span>table<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">1</span></span></code></pre></div>
<ul>
<li>That’s shocking. One one significant gene: <code>NR_003255|TSIX</code>. Let’s do one more pairwise comparison.</li>
</ul>
<p><br></p>
<h3 id="gbr-female-vs-yri-female">9.2 GBR Female vs YRI Female</h3>
<ul>
<li>These would be for groups 1 and 3.</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true"></a>GBRF_YRIF &lt;-<span class="st"> </span><span class="kw">exactTest</span>(ydisp, <span class="dt">pair=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true"></a>GBRF_YRIF</span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true"></a>GBRF_YRIF_FDR &lt;-<span class="st"> </span><span class="kw">topTags</span>(GBRF_YRIF, <span class="dt">n=</span><span class="ot">Inf</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true"></a><span class="kw">sum</span>(GBRF_YRIF_FDR<span class="op">$</span>table<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">0</span></span></code></pre></div>
<ul>
<li>It seems that there are no significant DE genes in this pair either. This may indicate problems with the analysis upstream of what we have done today. Just to be sure, let’s try one more pair.</li>
</ul>
<p><br></p>
<h3 id="gbr-male-vs-yri-female">9.3 GBR Male vs YRI Female</h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true"></a>GBRM_YRIF &lt;-<span class="st"> </span><span class="kw">exactTest</span>(ydisp, <span class="dt">pair=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true"></a>GBRM_YRIF</span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true"></a>GBRM_YRIF_FDR &lt;-<span class="st"> </span><span class="kw">topTags</span>(GBRM_YRIF, <span class="dt">n=</span><span class="ot">Inf</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true"></a><span class="kw">sum</span>(GBRM_YRIF_FDR<span class="op">$</span>table<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">6</span></span></code></pre></div>
<ul>
<li>There are six genes in this comparison that are significant. What are they?</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true"></a><span class="kw">subset</span>(GBRM_YRIF_FDR<span class="op">$</span>table, FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true"></a>                       logFC    logCPM       PValue         FDR</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true"></a>NR_<span class="dv">003255</span><span class="op">|</span>TSIX      <span class="fl">6.274395</span>  <span class="fl">8.731971</span> <span class="fl">4.195616e-06</span> <span class="fl">0.002706173</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true"></a>NR_<span class="dv">026595</span><span class="op">|</span>FAM226A   <span class="fl">2.404004</span>  <span class="fl">4.684012</span> <span class="fl">2.161911e-05</span> <span class="fl">0.004897595</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true"></a>NM_<span class="dv">004187</span><span class="op">|</span>KDM5C     <span class="fl">1.232794</span> <span class="fl">10.926060</span> <span class="fl">2.277951e-05</span> <span class="fl">0.004897595</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true"></a>NM_<span class="dv">001306145</span><span class="op">|</span>MTMR1 <span class="fl">-3.476830</span>  <span class="fl">4.682143</span> <span class="fl">5.607137e-05</span> <span class="fl">0.009041509</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true"></a>NM_<span class="dv">018390</span><span class="op">|</span>PLCXD1    <span class="fl">1.486596</span> <span class="fl">11.132709</span> <span class="fl">1.025941e-04</span> <span class="fl">0.013234645</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true"></a>NR_<span class="dv">109776</span><span class="op">|</span>ARHGAP6   <span class="fl">2.798704</span>  <span class="fl">2.504785</span> <span class="fl">1.807483e-04</span> <span class="fl">0.019430437</span></span></code></pre></div>
<p><br></p>
<h3 id="gbr-male-vs-yri-male">9.4 GBR Male vs YRI Male</h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true"></a>GBRM_YRIM &lt;-<span class="st"> </span><span class="kw">exactTest</span>(ydisp, <span class="dt">pair=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>))</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true"></a>GBRM_YRIM</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true"></a>GBRM_YRIM_FDR &lt;-<span class="st"> </span><span class="kw">topTags</span>(GBRM_YRIM, <span class="dt">n=</span><span class="ot">Inf</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true"></a><span class="kw">sum</span>(GBRM_YRIM_FDR<span class="op">$</span>table<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">0</span></span></code></pre></div>
<p><br></p>
<h3 id="gbr-female-vs-yri-male">9.5 GBR Female vs YRI Male</h3>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true"></a>GBRF_YRIM &lt;-<span class="st"> </span><span class="kw">exactTest</span>(ydisp, <span class="dt">pair=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true"></a></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true"></a>GBRF_YRIM</span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true"></a>GBRF_YRIM_FDR &lt;-<span class="st"> </span><span class="kw">topTags</span>(GBRF_YRIM, <span class="dt">n=</span><span class="ot">Inf</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true"></a><span class="kw">sum</span>(GBRF_YRIM_FDR<span class="op">$</span>table<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">0</span></span></code></pre></div>
<p><br></p>
<h3 id="yri-female-vs-yri-male">9.6 YRI Female vs YRI Male</h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true"></a>YRIF_YRIM &lt;-<span class="st"> </span><span class="kw">exactTest</span>(ydisp, <span class="dt">pair=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true"></a>YRIF_YRIM</span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true"></a>YRIF_YRIM_FDR &lt;-<span class="st"> </span><span class="kw">topTags</span>(YRIF_YRIM, <span class="dt">n=</span><span class="ot">Inf</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true"></a><span class="kw">sum</span>(YRIF_YRIM_FDR<span class="op">$</span>table<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="dv">0</span></span></code></pre></div>
<p><br><br></p>
<h2 id="subset-all-comparisons-for-the-6-significant-de-genes">10. Subset All Comparisons for the 6 Significant DE Genes</h2>
<ul>
<li>First create a simple vector containing names of those 6 genes.</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true"></a>genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NR_003255|TSIX&quot;</span>, <span class="st">&quot;NR_026595|FAM226A&quot;</span>, <span class="st">&quot;NM_004187|KDM5C&quot;</span>,</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true"></a><span class="st">&quot;NM_001306145|MTMR1&quot;</span>, <span class="st">&quot;NM_018390|PLCXD1&quot;</span>, <span class="st">&quot;NR_109776|ARHGAP6&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true"></a>sub1 &lt;-<span class="st"> </span><span class="kw">subset</span>(GBR_MF_FDR<span class="op">$</span>table, <span class="kw">rownames</span>(GBR_MF_FDR<span class="op">$</span>table) <span class="op">%in%</span><span class="st"> </span>genes)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true"></a>sub2 &lt;-<span class="st"> </span><span class="kw">subset</span>(GBRF_YRIF_FDR<span class="op">$</span>table, <span class="kw">rownames</span>(GBRF_YRIF_FDR<span class="op">$</span>table) <span class="op">%in%</span><span class="st"> </span>genes)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true"></a>sub3 &lt;-<span class="st"> </span><span class="kw">subset</span>(GBRM_YRIF_FDR<span class="op">$</span>table, <span class="kw">rownames</span>(GBRM_YRIF_FDR<span class="op">$</span>table) <span class="op">%in%</span><span class="st"> </span>genes)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true"></a>sub4 &lt;-<span class="st"> </span><span class="kw">subset</span>(GBRM_YRIM_FDR<span class="op">$</span>table, <span class="kw">rownames</span>(GBRM_YRIM_FDR<span class="op">$</span>table) <span class="op">%in%</span><span class="st"> </span>genes)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true"></a>sub5 &lt;-<span class="st"> </span><span class="kw">subset</span>(GBRF_YRIM_FDR<span class="op">$</span>table, <span class="kw">rownames</span>(GBRF_YRIM_FDR<span class="op">$</span>table) <span class="op">%in%</span><span class="st"> </span>genes)</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true"></a>sub6 &lt;-<span class="st"> </span><span class="kw">subset</span>(YRIF_YRIM_FDR<span class="op">$</span>table, <span class="kw">rownames</span>(YRIF_YRIM_FDR<span class="op">$</span>table) <span class="op">%in%</span><span class="st"> </span>genes)</span></code></pre></div>
<ul>
<li>Second, extract the logFC column from each and put into a single data frame.</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true"></a>logFC &lt;-<span class="st"> </span><span class="kw">data.frame</span>(sub1<span class="op">$</span>logFC, sub2<span class="op">$</span>logFC, sub3<span class="op">$</span>logFC, sub4<span class="op">$</span>logFC, sub5<span class="op">$</span>logFC, sub6<span class="op">$</span>logFC)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true"></a><span class="kw">rownames</span>(logFC) &lt;-<span class="st"> </span>genes</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true"></a>logFC</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true"></a>                   sub1.logFC sub2.logFC sub3.logFC sub4.logFC sub5.logFC sub6.logFC</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true"></a>NR_<span class="dv">003255</span><span class="op">|</span>TSIX     <span class="fl">-5.3282667</span> <span class="fl">-3.0164320</span>   <span class="fl">6.274395</span>  <span class="fl">2.5758704</span>  <span class="fl">2.5758704</span>  <span class="fl">2.5758704</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true"></a>NR_<span class="dv">026595</span><span class="op">|</span>FAM226A  <span class="fl">-0.6818842</span>  <span class="fl">2.1898706</span>   <span class="fl">2.404004</span>  <span class="fl">1.1532322</span>  <span class="fl">1.1532322</span>  <span class="fl">1.1532322</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true"></a>NM_<span class="dv">004187</span><span class="op">|</span>KDM5C    <span class="fl">-0.5414571</span>  <span class="fl">1.5945215</span>   <span class="fl">1.232794</span>  <span class="fl">1.3553918</span>  <span class="fl">1.3553918</span>  <span class="fl">1.3553918</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true"></a>NM_<span class="dv">001306145</span><span class="op">|</span>MTMR1 <span class="fl">-0.8092117</span>  <span class="fl">0.9451367</span>  <span class="fl">-3.476830</span> <span class="fl">-1.7765739</span> <span class="fl">-1.7765739</span> <span class="fl">-1.7765739</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true"></a>NM_<span class="dv">018390</span><span class="op">|</span>PLCXD1    <span class="fl">0.4589596</span>  <span class="fl">0.5509091</span>   <span class="fl">1.486596</span>  <span class="fl">0.5849666</span>  <span class="fl">0.5849666</span>  <span class="fl">0.5849666</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true"></a>NR_<span class="dv">109776</span><span class="op">|</span>ARHGAP6  <span class="fl">-0.6058542</span>  <span class="fl">0.9461587</span>   <span class="fl">2.798704</span>  <span class="fl">0.4847461</span>  <span class="fl">0.4847461</span>  <span class="fl">0.4847461</span></span></code></pre></div>
<p><br><br></p>
<h2 id="make-a-heatmap-of-logfc">11. Make A Heatmap of logFC</h2>
<ul>
<li>Convert logFC data frame to a matrix</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true"></a>logFCm &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(logFC)</span></code></pre></div>
<ul>
<li>Draw a heat map</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">4</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">5</span>))</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true"></a><span class="kw">pdf</span>(<span class="st">&quot;heatmap_chrX.pdf&quot;</span>, <span class="dt">width=</span><span class="dv">10</span>, <span class="dt">height=</span><span class="dv">10</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true"></a><span class="kw">heatmap</span>(logFCm)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true"></a><span class="kw">dev.off</span>()</span></code></pre></div>
<center>
<img src="heatmap_chrX.png" width=500>
</center>
<p><br><br></p>
<h2 id="homework">12. Homework</h2>
<ol type="1">
<li><p>Make a Github repository named <code>rnaseq_chrX</code></p></li>
<li><p>Recreate the heat map figure using <code>heatmap.2()</code> function in R</p></li>
<li><p>Write a short paragraph explaining what this heat map is showing or not showing</p></li>
<li><p>Create a new markdown file with this paragraph and the heatmap figure.</p></li>
<li><p>Push it to your GitHub repository</p></li>
<li><p>Post GitHub repo URL to #livecoding channel on Slack</p></li>
</ol>
<p><br><br><br><br> <br><br><br><br> <br><br><br><br> <br><br><br><br></p>
</body>
</html>
